corehq/ex-submodules/casexml/apps/case/mock/case_block.py

```diff
@@ -2,145 +2,167 @@
-            case_name = "" if case_name is CaseBlock.undefined else case_name
-            owner_id = "" if owner_id is CaseBlock.undefined else owner_id
-        self['update'] = update
-
-        create_or_update = {
-            self.CASE_TYPE: case_type,
-            'case_name': case_name,
-        }
+        self.date_modified = date_modified or now
+        self.date_opened = (now.date() if create and date_opened is CaseBlock.undefined
+                            else date_opened)
+
+        self.case_type = "" if create and case_type is CaseBlock.undefined else case_type
+        self.case_name = "" if create and case_name is CaseBlock.undefined else case_name
+        self.owner_id = "" if create and owner_id is CaseBlock.undefined else owner_id
+        self.close = close
+        self.case_id = case_id
+        self.user_id = user_id
+        self.external_id = external_id
+        self.create = create
+        self._check_for_duplicate_properties()
+        self.index = {key: self._make_index_attrs(value)
+                      for key, value in index.items()} if index else {}
+
+    def _updatable_built_ins(self):
+        return (('case_type', self.case_type),
+                ('case_name', self.case_name),
+                ('owner_id', self.owner_id))
+
+    def _check_for_duplicate_properties(self):
+        for property_name, passed_value in self._updatable_built_ins():
+            if passed_value is not CaseBlock.undefined and property_name in self.update:
+                    raise CaseBlockError("Key {!r} specified twice".format(property_name))
+
+    @staticmethod
+    def _make_index_attrs(value):
+        if len(value) == 2:
+            return IndexAttrs(value[0], value[1], 'child')
+        else:
+            case_type, case_id, relationship = value
+            if relationship not in ('child', 'extension'):
+                raise CaseBlockError('Valid values for an index relationship are "child" and "extension"')
+            return IndexAttrs(case_type, case_id, relationship)
 
-        self.update({
+    def _to_json(self):
+        result = {
             '_attrib': {
-                'case_id': case_id,
-                'date_modified': date_modified,
-                'user_id': user_id,
-                'xmlns': self.XMLNS,
-            }
-        })
-        if owner_id is not None:
-            create_or_update.update({
-                'owner_id': owner_id,
-            })
-        self['update'].update({
-            'external_id': external_id,
-            'date_opened': date_opened,
-        })
+                'case_id': self.case_id,
+                'date_modified': self.date_modified,
+                'user_id': self.user_id,
+                'xmlns': V2_NAMESPACE,
+            },
+            'update': self.update,
+        }
 
-        # fail if user specifies both, say, case_name='Johnny' and update={'case_name': 'Johnny'}
-        if strict:
```

Comment:
Why is the strict option removed? (I'm not fully through the PR, might become clear later on, but I think I've read the relevant code.)